File: servers/typescript/src/handler_rtc.ts
import type { StarfishFrame } from "./types.js";
import { parseTo } from "./types.js";
import type { Client } from "./client.js";
import type { StarfishServer } from "./starfish_server.js";
import {
  createErrorFrame,
  ERR_PROTOCOL_INVALID_FRAME,
  ERR_SESSION_NOT_FOUND,
  ERR_CLIENT_NOT_FOUND,
} from "./errors.js";

export function handleRTCConnect(hub: StarfishServer, client: Client, frame: StarfishFrame): void {
  relayRTC(hub, client, frame);
}

export function handleRTCOffer(hub: StarfishServer, client: Client, frame: StarfishFrame): void {
  relayRTC(hub, client, frame);
}

export function handleRTCAnswer(hub: StarfishServer, client: Client, frame: StarfishFrame): void {
  relayRTC(hub, client, frame);
}

export function handleRTCIce(hub: StarfishServer, client: Client, frame: StarfishFrame): void {
  relayRTC(hub, client, frame);
}

function relayRTC(hub: StarfishServer, client: Client, frame: StarfishFrame): void {
  const targets = parseTo(frame.header.to);
  if (targets.length !== 1) {
    client.sendFrame(
      createErrorFrame(hub.idGen, frame.header.id, ERR_PROTOCOL_INVALID_FRAME, "rtc", frame.header.method),
    );
    return;
  }

  const targetId = targets[0];

  const session = hub.getSession(frame.header.session!);
  if (!session) {
    client.sendFrame(
      createErrorFrame(hub.idGen, frame.header.id, ERR_SESSION_NOT_FOUND, "rtc", frame.header.method),
    );
    return;
  }

  const target = session.getClient(targetId);
  if (!target) {
    client.sendFrame(
      createErrorFrame(hub.idGen, frame.header.id, ERR_CLIENT_NOT_FOUND, "rtc", frame.header.method),
    );
    return;
  }

  target.sendFrame({
    header: {
      id: frame.header.id,
      resource: "rtc",
      method: frame.header.method,
      kind: "event",
      session: frame.header.session,
      from: client.id,
      to: targetId,
    },
    payload: frame.payload,
  });
}

File: servers/python/src/starfish_server/handler_rtc.py
from __future__ import annotations

from typing import TYPE_CHECKING

from .client import Client
from .errors import ERR_CLIENT_NOT_FOUND, ERR_PROTOCOL_INVALID_FRAME, ERR_SESSION_NOT_FOUND, create_error_frame
from .types import StarfishFrame, parse_to

if TYPE_CHECKING:
    from .server import StarfishServer


def handle_rtc_connect(hub: StarfishServer, client: Client, frame: StarfishFrame) -> None:
    _relay_rtc(hub, client, frame)


def handle_rtc_offer(hub: StarfishServer, client: Client, frame: StarfishFrame) -> None:
    _relay_rtc(hub, client, frame)


def handle_rtc_answer(hub: StarfishServer, client: Client, frame: StarfishFrame) -> None:
    _relay_rtc(hub, client, frame)


def handle_rtc_ice(hub: StarfishServer, client: Client, frame: StarfishFrame) -> None:
    _relay_rtc(hub, client, frame)


def _relay_rtc(hub: StarfishServer, client: Client, frame: StarfishFrame) -> None:
    header = frame.get("header", {})
    targets = parse_to(header.get("to"))

    if len(targets) != 1:
        client.send_frame(
            create_error_frame(hub.id_gen, header.get("id", ""), ERR_PROTOCOL_INVALID_FRAME, "rtc", header.get("method"))
        )
        return

    target_id = targets[0]
    session_name = header.get("session")

    session = hub.get_session(session_name) if session_name else None
    if session is None:
        client.send_frame(
            create_error_frame(hub.id_gen, header.get("id", ""), ERR_SESSION_NOT_FOUND, "rtc", header.get("method"))
        )
        return

    target = session.get_client(target_id)
    if target is None:
        client.send_frame(
            create_error_frame(hub.id_gen, header.get("id", ""), ERR_CLIENT_NOT_FOUND, "rtc", header.get("method"))
        )
        return

    target.send_frame({
        "header": {
            "id": header.get("id", ""),
            "resource": "rtc",
            "method": header.get("method"),
            "kind": "event",
            "session": session_name,
            "from": client.id,
            "to": target_id,
        },
        "payload": frame.get("payload"),
    })
